### ISSUE TITLE
XML/DOM: Remove raw pointers from public interfaces

### ISSUE THREAD

**Issue by obiltschnig** [pocoproject/poco#2075]
Raw pointers in XML/DOM interfaces should be replaced with smart pointers (`Poco::AutoPtr`).

Unfortunately, this cannot be done without breaking existing code. After extensive analysis, a method has been found to implement this with minimal backwards compatibility issues. The main concern was that code requiring changes should not break more or less silently or crash at runtime (e.g., creating memory leaks or double deletes), but break at compile time.

For passing DOM objects as method arguments and return values, the following mechanism will be used:

  1. The `duplicate()` and `release()` methods in `DOMObject` will be made `protected`, therefore     preventing explicit calls to them from client code. This will require changes to client code not already using `Poco::AutoPtr` (as it should already).

  2. A new class template `Poco::XML::NodePtr` will be created for passing DOM Node objects as method arguments. This will be a subclass of `Poco::AutoPtr`, but with the semantics of the constructor taking a raw pointer changed to call duplicate(). This way, existing code using raw pointers will continue to work with correct reference counting. However, this constructor will be marked as deprecated, so that a warning will be issued when it's being used.

  3. All methods returning raw `DOMObject` pointers will be changed to return `Poco::AutoPtr`. This should not break any code, except code that explicitly calls `release()`, e.g., after receiving a `NodeList` or `NamedNodeMap`. However, since `release()` is no longer public, breakage will occur during compilation, and not during runtime in the form of memory leaks or double deletes.

  4. Eventually, in a later version, `Poco::XML::NodePtr` will be removed and replaced with `Poco::AutoPtr`.

**Comment by dawidpilarski** [comment]
Any reason, why this  issue was marked as wontfix? 

I am willing to contribute to that particular issue as it already caused lots of troubles in one of the projects I am working on.

**Comment by obiltschnig** [comment]
No way to do this in a way that keeps it backwards compatible to existing code. So basically we'll have to implement a new API in parallel to the existing one. There's simply no resources to do it.

**Comment by dawidpilarski** [comment]
Thanks for quick response @obiltschnig 

Are changes breaking the code at compile time (as previously considered) also not an option?

> The main concern was that code requiring changes should not break more or less silently or crash at runtime (e.g., creating memory leaks or double deletes), but break at compile time.

I think such an approach would be possible.

**Comment by obiltschnig** [comment]
I'd like to try to avoid that. Of course breaking at compile time is preferable to breaking at run-time, but I can imagine such a breaking change not going well with most users - everyone usually has enough work to do and no time to rewrite code just because the framework changes. Maybe deprecating the existing DOM APIs and introducing a new library with AutoPtr-based interfaces would be an option. And while at it - there are a few other things in the DOM library that need improvement (the `NodeList` implementations are a bit of a disaster ;-)).

**Comment by obiltschnig** [comment]
If you're interested in doing this, we would much appreciate it, of course.
